pkg/controller/constraint/constraint_controller.go

```diff
@@ -128,6 +163,11 @@ func (r *ReconcileConstraint) Reconcile(request reconcile.Request) (reconcile.Re
 		util.SetHAStatus(instance, status)
 
 		if _, err := r.opa.AddConstraint(context.Background(), instance); err != nil {
+			r.constraintsCache[constraintKindName] = Tags{
```

Comment:
If we're using read locks below, we should use write locks here.

Generally speaking, access to the underlying data structure should be gated by accessor methods which implement the locking/unlocking. This has the side benefit of helping ensure you haven't accidentally deadlocked as locks are released before the function returns, making it a little easier to verify the code.

Reply:
changed to write locks and moved populating constraintCache inside. wdyt?

Reply:
LGTM